src/wrapper/clap-juce-wrapper.cpp

```diff
@@ -819,176 +824,219 @@ class ClapJuceWrapper : public clap::helpers::Plugin<
-            }
+            auto event = events->get(events, 0);
+            nextEventTime = (int)event->time;
         }
 
-        // We process in place so
-        static constexpr uint32_t maxBuses = 128;
-        std::array<float *, maxBuses> busses{};
-        busses.fill(nullptr);
+        auto processEvent = [&](int sampleOffset) {
+            auto event = events->get(events, (uint32_t)currentEvent);
+            process_clap_event(event, sampleOffset);
 
-        /*DBG("IO Configuration: I=" << (int)process->audio_inputs_count << " O="
-                                   << (int)process->audio_outputs_count << " MX=" << (int)mx);
-        DBG("Plugin Configuration: IC=" << processor->getTotalNumInputChannels()
-                                        << " OC=" << processor->getTotalNumOutputChannels());
-        */
+            currentEvent++;
+            nextEventTime = (currentEvent < numEvents)
+                                ? (int)events->get(events, (uint32_t)currentEvent)->time
+                                : numSamples;
+        };
 
         /*
          * OK so here is what JUCE expects in its audio buffer. It *always* uses input as output
          * buffer so we need to create a buffer where each channel is the channel of the associated
          * output pointer (fine) and then the inputs need to either check they are the same or copy.
          */
+        static constexpr uint32_t maxBuses = 128;
+        std::array<float *, maxBuses> busses{};
+        busses.fill(nullptr);
 
-        /*
-         * So first lets load up with our outputs
-         */
-        uint32_t ochans = 0;
-        for (uint32_t idx = 0; idx < process->audio_outputs_count && ochans < maxBuses; ++idx)
+        for (int n = 0; n < numSamples;)
         {
-            for (uint32_t ch = 0; ch < process->audio_outputs[idx].channel_count; ++ch)
+            while (nextEventTime == n) // process all events at this timestep
+                processEvent(n);
+
+            auto getSamplesToProcess = [&]() {
+                if (CLAP_EVENT_RESOLUTION_SAMPLES <= 0)
+                {
+                    // Sample-accurate events are turned off, so just process the whole block!
+                    return numSamples;
+                }
+                else
+                {
+                    // How many samples should we process at a time?
+                    // In the spirit of sample-accurate events, we want to process a batch of
+                    // samples until we hit the next event, but we don't want to have a batch
+                    // smaller than the `CLAP_EVENT_RESOLUTION_SAMPLES`. If there's no more
+                    // events, just process the rest of the block!
+                    return (numSamples - n >= CLAP_EVENT_RESOLUTION_SAMPLES)
+                               ? juce::jmax(nextEventTime - n, CLAP_EVENT_RESOLUTION_SAMPLES)
+                               : (numSamples - n);
+                }
+            };
+
```

Comment:
So this is going to make some very odd sized blocks. I wonder if we want to quantize somehow. While most juce plugins will handle blocks of size 3 or whatever, its a bit antisocial it seems. For surge effects it will result in us having to always run with latency on for instance, even with systems which give us %32 sized and fixed blocks.

While this absolutely accurate mode makes sense I do wonder if you want to allow a preferred or best efforts subblock floor.

Reply:
That's a good point... how about this instead:
```cpp
auto getSamplesToProcess = [&]() {
    if (CLAP_EVENT_RESOLUTION_SAMPLES <= 0)
    {
        // Sample-accurate events are turned off, so just process the
        // whole block.
        return numSamples;
    }
    else
    {
        // the number of samples left is less than CLAP_EVENT_RESOLUTION_SAMPLES
        // so let's just process the rest of the block
        if (numSamples - n < CLAP_EVENT_RESOLUTION_SAMPLES)
            return numSamples - n;

        // process up until the next event, rounding up to the nearest multiple
        // of CLAP_EVENT_RESOLUTION_SAMPLES
        return ((nextEventTime - n) + CLAP_EVENT_RESOLUTION_SAMPLES - 1) / CLAP_EVENT_RESOLUTION_SAMPLES;
    }
};
```

Reply:
So we want to be careful. A mid block midi event for instance is sent with a time stamp into process so this will break note streams pointlessly - we definitely have to have some event types which don’t split us. Midi and note on and the like.

I also kinda was thinking - we just always do 32 no matter what (or multiples thereof) might be useful mode.

Reply:
Gotcha... so basically, we don't want to split a block for an event, if the event is of type `CLAP_EVENT_NOTE_ON`, `CLAP_EVENT_NOTE_OFF`, or `CLAP_EVENT_MIDI`? Are there any event types I'm missing there (maybe note expression events?).

I don't love the idea of always splitting the block, since it seems wasteful for situations where there is no parameter automation, or where the host event resolution is coarser than the plugin event resolution, but I guess it couldn't hurt to have as an option.